service/storage_proxy.cc

```diff
@@ -3515,6 +3547,9 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
     }
     future<> make_digest_requests(digest_resolver_ptr resolver, targets_iterator begin, targets_iterator end, clock_type::time_point timeout) {
         return parallel_for_each(begin, end, [this, resolver = std::move(resolver), timeout] (gms::inet_address ep) {
+            if (maybe_speculatively_fail(*resolver, ep, timeout)) {
+                return make_ready_future<>();
+            }
```

Comment:
Shouldn't we pick another replica instead?

Reply:
And mark this as an error?

Reply:
It's marked as an error by putting the error in the resolver, inside `maybe_speculatively_fail`. Returning an exception_future here would just result in a warning that an exceptional future is ignored, since everything is waited for on the resolver.

As for picking replicas - speculative retry would do that for free if it's enabled and the resolver would not receive enough acks to fulfill the CL. If speculative retry is not enabled, then perhaps we can indeed try being smarter and suggesting another replica, although I expect the patch to be rather intrusive.

Reply:
So are we re-inventing dynamic snitch?

Reply:
I agree with @gleb-cloudius that we should have answered this question before showing the code. I tried to do it in a new issue https://github.com/scylladb/scylla/issues/8964 - I hope @psarna agrees with my description of his idea, and my explanation of why this idea is different (and better) than the dynamic snitch will please @gleb-cloudius.